File: lib/rgfa/field/alignment_list_gfa1.rb
module RGFA::Field::AlignmentListGFA1

  def unsafe_decode(string)
    string.split(",").map {|c| c.to_cigar(valid: true, version: :gfa1)}
  end

  def decode(string)
    validate_encoded(string)
    unsafe_decode(string)
  end

  def validate_encoded(string)
     if string !~ /^(\*|(([0-9]+[MIDNSHPX=])+))(,(\*|(([0-9]+[MIDNSHPX=])+)))*$/
       raise RGFA::FormatError,
         "#{string.inspect} is not a comma separated list of * or CIGARs\n"+
         "(CIGAR strings must match ([0-9]+[MIDNSHPX=])+)"
     end
  end

  def validate_decoded(object)
    case object
    when RGFA::Placeholder
    when Array
      object.map do |elem|
        elem.to_cigar(version: :gfa1)
      end.each(&:validate)
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, RGFA::Placeholder)"
    end
  end

  def unsafe_encode(object)
    case object
    when RGFA::Placeholder
      object.to_s
    when Array
      object.map{|cig|cig.to_cigar.to_s}.join(",")
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, RGFA::Placeholder)"
    end
  end

  def encode(object)
    case object
    when RGFA::Placeholder
      object.to_s
    when Array
      object.map do |cig|
        cig = cig.to_cigar
        cig.validate
        cig.to_s
      end.join(",")
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: Array, RGFA::Placeholder)"
    end
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/alignment_list_gfa1.py
import gfapy
import re

def unsafe_decode(string):
  return [ gfapy.Alignment(s, version = "gfa1", valid = True) \
             for s in string.split(",") ]

def decode(string):
  validate_encoded(string)
  return unsafe_decode(string)

def validate_encoded(string):
  if not re.match(r"^(\*|(([0-9]+[MIDNSHPX=])+))(,(\*|(([0-9]+[MIDNSHPX=])+)))*$", string):
     raise gfapy.FormatError(
       "{} is not a comma separated list of * or CIGARs\n".format(repr(string))+
       "(CIGAR strings must match ([0-9]+[MIDNSHPX=])+)")

def validate_decoded(obj):
  if isinstance(obj, gfapy.Placeholder):
    pass
  elif isinstance(obj, list):
    for e in obj:
      e = gfapy.Alignment(e, version = "gfa1")
      e.validate()
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n".format(obj.__class__.__name__)+
      "(accepted classes: list, AlignmentPlaceholder)")

def unsafe_encode(obj):
  if isinstance(obj, gfapy.Placeholder):
    return str(obj)
  elif isinstance(obj, list):
    return ",".join([str(gfapy.Alignment(cig)) for cig in obj])
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: list, AlignmentPlaceholder)")

def encode(obj):
  if isinstance(obj, gfapy.Placeholder):
    return str(obj)
  if isinstance(obj, list):
    def f(cig):
      cig = gfapy.Alignment(cig)
      cig.validate()
      return str(cig)
    return ",".join(map(f, obj))
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: list, AlignmentPlaceholder)")
